### PR TITLE
Set `org.gradle.plugin.api-version` in buildSrc and `java-gradle-plugin`

### LINKED ISSUE #20847: Plugin variants are ignored when plugin is added as buildSrc common dependency
### Expected Behavior

When plugin with variants is added as a common dependency in `buildSrc` module, related variant is selected:
```kt
plugins {
   `kotlin-dsl`
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.0-RC"
}
```

And in project with Gradle version 7+ - Gradle should select `gradle70` variant.

### Current Behavior

Gradle selects main plugin variant which doesn't have `GRADLE_PLUGIN_API_VERSION_ATTRIBUTE` attribute.

It is possible to workaround by adding such attribute to all resolvable configurations:
```kt
configurations.configureEach {
    if (isCanBeResolved) {
        attributes {
            @Suppress("UnstableApiUsage")
            attribute(GradlePluginApiVersion.GRADLE_PLUGIN_API_VERSION_ATTRIBUTE, project.objects.named(GradleVersion.current().version))
        }
    }
}
```

### Context

User may assume build is using latest plugin variant while it is not. If plugin main variant has only basic functionality - it could lead to build performance regression.

### Steps to Reproduce 

- Create a basic project with `buildSrc` and a code above and latest Gradle version
- check info logs for `Using Kotlin Gradle plugin gradle70 variant` line, while actually it will be `Using Kotlin Gradle plugin main variant`

### PR DESCRIPTION
**Pull request by h0tk3y**
Fixes #20847

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/developing-plugins/implementing_gradle_plugins.adoc [hunk 1]
```diff
@@ -352,3 +352,25 @@ This will be improved in the future.
...
+* plugins specified in the <<plugins.adoc#sec:plugins_block,`plugins {}` block>>;
+* `buildscript` classpath dependencies;
+* dependencies in the root project of the <<organizing_gradle_projects.adoc#sec:build_sources,build source (`buildSrc`)>> that appear on the compile or runtime classpath;
+* dependencies in a project that applies the <<java_gradle_plugin.adoc#java_gradle_plugin,Java Gradle Plugin Development plugin>> or the <<kotlin_dsl.adoc#sec:kotlin-dsl_plugin,Kotlin DSL plugin>>, appearing on the compile or runtime classpath.
+
+The best matching variant is the variant that targets the highest Gradle API version not exceeding the current build's Gradle version.
+
+In all other cases, a plugin variant that does not specify the supported Gradle API version is preferred, if such a variant is present.
+
+In projects that use plugins as dependencies, it is possible to request the variants of plugin dependencies that support a different Gradle version. This allows a
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/developing-plugins/implementing_gradle_plugins.adoc:368]: Doc style: one sentence per line

**Review by ljacomet — APPROVED** [review]
LGTM!

One minor doc comment, go ahead and merge once resolved.

File: subprojects/docs/src/docs/userguide/developing-plugins/implementing_gradle_plugins.adoc (hunk 1 above)
Comment (h0tk3y) [reply subprojects/docs/src/docs/userguide/developing-plugins/implementing_gradle_plugins.adoc]: Fixed, thanks!

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Pull request merged**